ecid_signal_measurement_info lwm2m-object (ID 10256)

jarlamsa opened the pull request:
Provide support for the ecid_signal_measurement_info lwm2m-object (ID 10256)
The aim of the object is to hold the information about the neighbouring cells.
The information is then used for cell location measurement.

The library provides the creation and runtime modification of the objects and handling the neighbour cell measurement coming from the modem. The application needs to trigger the neighbour cell measurements itself.

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-6600
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

VeijoPesonen commented on subsys/net/lib/lwm2m_client_utils/CMakeLists.txt:
```diff
@@ -13,3 +13,5 @@ zephyr_library_sources_ifdef(CONFIG_LWM2M_CLIENT_UTILS_FIRMWARE_UPDATE_OBJ_SUPPO
 zephyr_library_sources_ifdef(CONFIG_LWM2M_CLIENT_UTILS_CONN_MON_OBJ_SUPPORT lwm2m/lwm2m_connmon.c)
 zephyr_library_sources_ifdef(CONFIG_LWM2M_CLIENT_UTILS_DEVICE_OBJ_SUPPORT lwm2m/lwm2m_device.c)
 zephyr_library_sources_ifdef(CONFIG_LWM2M_CLIENT_UTILS_SECURITY_OBJ_SUPPORT lwm2m/lwm2m_security.c)
+zephyr_library_sources_ifdef(CONFIG_LWM2M_CLIENT_UTILS_SIGNAL_MEAS_INFO_OBJ_SUPPORT lwm2m/ecid_signal_meas_info.c)
```
No need to check for the macro twice - for [example](https://github.com/zephyrproject-rtos/zephyr/blob/25e6803705b25602eb0b730a03a254c354bfae9e/subsys/net/lib/lwm2m/CMakeLists.txt#L33)

VeijoPesonen commented on samples/nrf9160/lwm2m_client/src/main.c:
```diff
@@ -544,6 +564,10 @@ void main(void)
 #endif
 
+#if defined(CONFIG_LWM2M_CLIENT_UTILS_SIGNAL_MEAS_INFO_OBJ_SUPPORT)
+	k_timer_start(&ncell_meas_work_timer, K_SECONDS(1), K_SECONDS(60));
```
Would it make sense to have these as Kconfig parameters?

jarlamsa replied:
Maybe the interval would be.

jarlamsa commented:
Thank you for your review @VeijoPesonen, I think I have addressed your comments, please re-review.

SeppoTakalo commented on subsys/net/lib/lwm2m_client_utils/Kconfig:
```diff
@@ -59,6 +59,19 @@ config LWM2M_CLIENT_UTILS_DOWNLOADER_SEC_TAG
+	help
+	  This object is used in combination with Connectivity Monitor object (ID 4)
+	  and provides information about neighbouring cells. It is used to assist
+	  with the device location services.
```
I think we are giving a way documentation that might not always be relevant.
We MAY use this for locationing purposes, however, it does not affect the code at all. Somebody else may use for some other purposes.
So should we document just what this functionality does here, which is providing the information regarding neighbor cells.

SeppoTakalo commented on subsys/net/lib/lwm2m_client_utils/lwm2m/neighbour_cell_info.c:
```diff
@@ -0,0 +1,107 @@
+	lwm2m_engine_set_s32(path, 0);
+}
+
+static void update_signal_meas_objects(struct lte_lc_cells_info cells)
```
Do you really want to receive copy of the struct in the parameter, or would it be less stack use if you would just bring the pointer?

SeppoTakalo commented on subsys/net/lib/lwm2m_client_utils/lwm2m/neighbour_cell_info.c:
```diff
@@ -0,0 +1,107 @@
+#define SIGNAL_MEAS_INFO_RSRQ_RESULT		4
+#define SIGNAL_MEAS_INFO_UE_RXTX_TIMEDIFF	5
+
+static void update_signal_meas_object(struct lte_lc_ncell cell, uint16_t index)
```
Do you really want to receive copy of the struct in the parameter, or would it be less stack use if you would just bring the pointer?

SeppoTakalo reviewed (approved):
LGTM,
I'm just wondering why you copy the whole struct, instead of using pointer in the parameter.

rlubos commented on subsys/net/lib/lwm2m_client_utils/lwm2m/ecid_signal_meas_info.c:
```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
```
I guess we could move to 2022 already :)

rlubos commented on subsys/net/lib/lwm2m_client_utils/lwm2m/ecid_signal_meas_info.c:
```diff
@@ -0,0 +1,119 @@
+
+#include "lwm2m_object.h"
+#include "lwm2m_engine.h"
+#include "net/lwm2m_client_utils.h"
```
~~This is actually a public header, so we should include it with square brackets <>.~~

Nevermind, I see it's done in the next commit.

rlubos commented on subsys/net/lib/lwm2m_client_utils/lwm2m/ecid_signal_meas_info.c:
```diff
@@ -59,6 +59,33 @@ static struct lwm2m_engine_obj_inst inst[MAX_INSTANCE_COUNT];
+		return -EINVAL;
+	}
+
+	/* not instanstiated */
```
nit: s/instanstiated/instantiated

rlubos commented on subsys/net/lib/lwm2m_client_utils/lwm2m/neighbour_cell_info.c:
```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
```
2022

rlubos commented on subsys/net/lib/lwm2m_client_utils/lwm2m/neighbour_cell_info.c:
```diff
@@ -0,0 +1,107 @@
+LOG_MODULE_REGISTER(lwm2m_neighbour_cell, CONFIG_LWM2M_CLIENT_UTILS_LOG_LEVEL);
+#define MAX_INSTANCE_COUNT CONFIG_LWM2M_CLIENT_UTILS_SIGNAL_MEAS_INFO_INSTANCE_COUNT
+
+#define SIGNAL_MEAS_INFO_PHYS_CELL_ID		0
+#define SIGNAL_MEAS_INFO_ECGI			1
+#define SIGNAL_MEAS_INFO_ARFCN_EUTRA		2
+#define SIGNAL_MEAS_INFO_RSRP_RESULT		3
+#define SIGNAL_MEAS_INFO_RSRQ_RESULT		4
+#define SIGNAL_MEAS_INFO_UE_RXTX_TIMEDIFF	5
```
I think it'd make sense to declare them in a header (along with object ID) instead of duplicating definitions here.

jarlamsa replied:
Good notion. Actually now that I think of it, these aren't even used here, only in the object source-file.

rlubos commented on samples/nrf9160/lwm2m_client/src/main.c:
```diff
@@ -108,6 +109,22 @@ static int query_modem(const char *cmd, char *buf, size_t buf_len)
 }
 
+#if defined(CONFIG_LWM2M_CLIENT_UTILS_SIGNAL_MEAS_INFO_OBJ_SUPPORT)
+void ncell_meas_work_handler(struct k_work *work)
+{
+	lte_lc_neighbor_cell_measurement(LTE_LC_NEIGHBOR_SEARCH_TYPE_DEFAULT);
+}
+
+K_WORK_DEFINE(ncell_meas_work, ncell_meas_work_handler);
+
+void ncell_meas_work_timer_handler(struct k_timer *dummy)
+{
+	k_work_submit(&ncell_meas_work);
+}
+
+K_TIMER_DEFINE(ncell_meas_work_timer, ncell_meas_work_timer_handler, NULL);
```
Instead of a timer/work combo I suggest simply using delayable work (https://docs.zephyrproject.org/latest/reference/kernel/threads/workqueue.html#delayable-work) - a work that is executed after a specific time.

rlubos reviewed (commented):
Some additional feedback, but looks pretty good overall.

jarlamsa commented:
Thank you for your reviews @SeppoTakalo @rlubos I think I have addressed your review comments. Please re-review.
I am also adding the changelog soon.

jarlamsa commented:
Rebased and added changelog

SeppoTakalo commented on samples/nrf9160/lwm2m_client/Kconfig:
```diff
@@ -166,6 +166,13 @@ config APP_MCUBOOT_FLASH_BUF_SZ
+config APP_NEIGHBOUR_CELL_SCAN_INTERVAL
+	int "Interval time [s] between neighbour cell scans"
+	depends on LWM2M_CLIENT_UTILS_SIGNAL_MEAS_INFO_OBJ_SUPPORT
+	default 60
```
Probably nitpicking, but should the default be a much more longer? Closer to 10 minutes?

SeppoTakalo commented on subsys/net/lib/lwm2m_client_utils/lwm2m/neighbour_cell_info.c:
```diff
@@ -0,0 +1,100 @@
+	lwm2m_engine_set_s32(path, 0);
+	snprintf(path, sizeof(path), "10256/%" PRIu16 "/4", obj_inst_id);
+	lwm2m_engine_set_s32(path, 0);
+	snprintf(path, sizeof(path), "10256/%" PRIu16 "/5", obj_inst_id);
```
Now I' not sure, but should we use `snprintk()` for modules?

umapraseeda commented on doc/nrf/releases/release-notes-changelog.rst:
```diff
@@ -239,6 +243,10 @@ Libraries for networking
 
+* :ref:`lib_lwm2m_client_utils` library:
+
+  * Added support for LwM2M-object ID 10256 (ECID-Signal Measurement Information).
```
```suggestion
  * Added support for LwM2M object ECID-Signal Measurement Information (object ID 10256).
```

rlubos reviewed (approved):
LGTM

jarlamsa commented:
@rlubos there seems to be some CI-failure that I don't think is related to this PR
```DEBUG   - DEVICE: Assertion failed at WEST_TOPDIR/nrf/tests/subsys/bootloader/bl_crypto/src/main.c:134: test_sha256_string: (test_vector[i] not equal to output[i])
DEBUG   - DEVICE: positive test failed (run no. 0) for i = 0
DEBUG   - DEVICE: FAIL - test_sha256 in 0.18 seconds```

Can we retrigger the CI somehow?

jarlamsa commented:
Rebased

jarlamsa commented:
Rebased
